sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-calculation/e-cert-calculation-process.ts

```diff
@@ -0,0 +1,140 @@
+import { ECertProcessStep } from "../e-cert-processing-steps/e-cert-steps-models";
+import { DataSource } from "typeorm";
+import { ProcessSummary } from "@sims/utilities/logger";
+import { parseJSONError, processInParallel } from "@sims/utilities";
+import { EligibleECertDisbursement } from "../disbursement-schedule.models";
+
+/**
+ * Disbursements grouped by student to allow parallel processing of students
+ * and sequential processing of disbursements for the same student.
+ */
+interface GroupedDisbursementPerStudent {
+  [studentId: number]: EligibleECertDisbursement[];
+}
+
+/**
+ * Base process for e-Cert calculations independently of the offering intensity.
+ * Both offering intensity will have similar calculation steps and some additional ones.
+ */
+export abstract class ECertCalculationProcess {
+  constructor(private readonly dataSource: DataSource) {}
+
+  /**
+   * Get all disbursements currently eligible to be part of
+   * an e-Cert to have its calculations executed.
+   * The returned array of {@link ECertProcessStep} will be shared across all steps being
+   * modified till the disbursement entity model modifications will be saved.
+   * */
+  protected abstract getDisbursements(): Promise<EligibleECertDisbursement[]>;
+
+  /**
+   * Define the steps to be executed and the execution order.
+   */
+  protected abstract calculationSteps(): ECertProcessStep[];
+
+  /**
+   * Execute the calculations for all eligible disbursements.
+   * @param log cumulative process log.
+   */
+  async executeCalculations(log: ProcessSummary): Promise<void> {
+    log.info("Retrieving eligible disbursements(s).");
+    const disbursements = await this.getDisbursements();
+    // Group retrieved disbursements per students since a student can
+    // potentially have more then one disbursement in the same e-Cert.
+    const disbursementsPerStudent =
+      this.getGroupedDisbursementsPerStudent(disbursements);
+    log.info(
+      `Found ${disbursements.length} eligible disbursement(s) for ${
+        Object.keys(disbursementsPerStudent).length
+      } student(s).`,
+    );
+    // Identify the distinct students to be processed in parallel.
+    const uniqueStudentIds = Object.keys(disbursementsPerStudent);
+    // Process students in parallel.
+    await processInParallel(
+      (uniqueStudentId: string) =>
+        this.calculateDisbursementsForStudent(
+          disbursementsPerStudent[uniqueStudentId],
+          log,
+        ),
+      uniqueStudentIds,
+    );
+  }
+
+  /**
+   * Group all disbursements per student id.
+   * @param eCertDisbursements disbursements to be grouped.
+   * @returns grouped disbursements.
+   */
+  private getGroupedDisbursementsPerStudent(
+    eCertDisbursements: EligibleECertDisbursement[],
+  ): GroupedDisbursementPerStudent {
+    return eCertDisbursements.reduce(
+      (
+        group: { [studentId: number]: EligibleECertDisbursement[] },
```

Comment:
we can use the interface `GroupedDisbursementPerStudent ` here

Reply:
Changed!